File: src/hanlint/rules/registry.py
"""규칙 등록부.

규칙 하나는 함수 하나다. `@rule("이름")` 으로 등록한다. 등록 시점에 셋을 거부한다. 이름이 함수 이름과
다른 것, docstring 이 없는 것, docstring 에 네 절 (왜, 어디서, 고치기, 안 잡는 것) 이 없는 것. 규칙의
docstring 은 기술서라 `hanlint explain` 이 그대로 보여 주고 네 절이 빠지면 기술서가 아니다.

규칙 모듈은 `loadAll` 이 부류 폴더를 걸어 전부 import 한다. 규칙 파일이 서로를 import 하지 않아도 되는
이유가 이것이다.
"""

from __future__ import annotations

import importlib
import inspect
import pkgutil
from collections.abc import Callable, Iterable

from ..config import Config
from ..fingerprint import DocumentPrint
from .finding import Finding

Check = Callable[[DocumentPrint, Config], Iterable[Finding]]
REGISTRY: dict[str, Check] = {}
REQUIRED_SECTIONS = ("왜:", "어디서:", "고치기:", "안 잡는 것:")
CATEGORIES = ("sentence", "paragraph", "structure", "document", "orthography", "code")
CATEGORY_TITLES = {
    "sentence": "문장 안에서 세는 것",
    "paragraph": "문단 사이에서 세는 것",
    "structure": "글의 짜임에서 세는 것",
    "document": "두 자리를 대조해 세는 것",
    "orthography": "표기와 띄어쓰기",
    "code": "코드 블록 사이를 대조하는 것",
}
"""부류의 사람 이름. `hanlint rules` 가 이 순서로 묶어 보인다. 뜻의 정본은 start.product 의 잡는 것이다."""
MECHANISMS = {
    "dictionary": "사전. 낱말 목록과 정규식이 맞는 자리",
    "repeat": "반복. 같은 모양이 창 안에서 N 번",
    "threshold": "셈. 지문의 값이 임계를 넘거나 모양이 계약과 다른 자리",
    "contrast": "대조. 두 자리를 맞대 어긋난 곳",
    "reader": "독자 상태. 문장 순서대로 독자가 손에 든 것과 본 것에 그 자리가 요구하는 것을 맞댄다",
}
"""규칙이 세는 방법의 닫힌 집합. 규칙은 쌓여도 기제는 늘지 않는다. 등록 시점에 이 밖의 기제를 거부하므로
여섯째 기제는 규칙 하나 때문에 조용히 들어오지 못한다. 새 기제가 정말 필요하면 여기를 고치기 전에 운영자에게
묻는다. 규칙과 기제의 대응은 `hanlint rules --format json` 과 npm 투영 `ruleMechanisms.json` 이 든다."""
MECHANISM_OF: dict[str, str] = {}


def rule(name: str, mechanism: str) -> Callable[[Check], Check]:
    if mechanism not in MECHANISMS:
        raise ValueError(
            f"규칙 {name} 의 기제 {mechanism} 은 닫힌 집합 밖이다 ({', '.join(MECHANISMS)}). "
            "새 기제는 규칙 하나 때문에 들어오지 않는다. 멈춰서 묻는다"
        )

    def register(check: Check) -> Check:
        if check.__name__ != name:
            raise ValueError(f"규칙 이름 {name} 과 함수 이름 {check.__name__} 이 다르다. 셋 (파일, 함수, id) 이 같아야 한다")
        doc = inspect.getdoc(check) or ""
        missing = [section for section in REQUIRED_SECTIONS if section not in doc]
        if not doc or missing:
            raise ValueError(f"규칙 {name} 의 docstring 에 {', '.join(missing) or '내용'} 이 없다. 기술서가 아니다")
        if name in REGISTRY and REGISTRY[name] is not check:
            raise ValueError(f"규칙 이름이 겹친다: {name}")
        REGISTRY[name] = check
        MECHANISM_OF[name] = mechanism
        return check

    return register


def loadAll() -> None:
    package = importlib.import_module("hanlint.rules")
    for category in CATEGORIES:
        module = importlib.import_module(f"hanlint.rules.{category}")
        for info in pkgutil.iter_modules(module.__path__, f"{module.__name__}."):
            importlib.import_module(info.name)
    del package


def ruleNames() -> list[str]:
    loadAll()
    return sorted(REGISTRY)


def ruleCategory(name: str) -> str:
    """규칙이 사는 부류 폴더 이름. 파일 위치가 정본이라 따로 적지 않는다."""
    loadAll()
    if name not in REGISTRY:
        raise KeyError(f"모르는 규칙: {name}. hanlint rules 로 목록을 본다")
    return REGISTRY[name].__module__.split(".")[-2]


def ruleCategories() -> dict[str, str]:
    """규칙 이름 → 부류. npm 투영 (`scripts/exportData.py`) 이 이것을 그대로 쓴다."""
    return {name: ruleCategory(name) for name in ruleNames()}


def ruleMechanism(name: str) -> str:
    """규칙이 세는 방법. MECHANISMS 의 키 하나."""
    loadAll()
    if name not in MECHANISM_OF:
        raise KeyError(f"모르는 규칙: {name}. hanlint rules 로 목록을 본다")
    return MECHANISM_OF[name]


def ruleMechanisms() -> dict[str, str]:
    """규칙 이름 → 기제. npm 투영 (`scripts/exportData.py`) 이 이것을 그대로 쓴다."""
    return {name: ruleMechanism(name) for name in ruleNames()}


def ruleDoc(name: str) -> str:
    loadAll()
    if name not in REGISTRY:
        raise KeyError(f"모르는 규칙: {name}. hanlint rules 로 목록을 본다")
    return inspect.getdoc(REGISTRY[name]) or ""


def ruleSummary(name: str) -> str:
    return ruleDoc(name).splitlines()[0]


def isDisabledAt(finding: Finding, disabled: tuple[tuple[str, int, int], ...]) -> bool:
    """인라인 제어가 그 줄에서 그 규칙을 껐는가. `*` 는 전부다."""
    return any((name == "*" or name == finding.rule) and start <= finding.line <= end for name, start, end in disabled)


def runAll(doc: DocumentPrint, config: Config) -> list[Finding]:
    loadAll()
    findings: list[Finding] = []
    for name in sorted(REGISTRY):
        if config.enabled(name):
            findings.extend(f for f in REGISTRY[name](doc, config) if not isDisabledAt(f, doc.disabled))
    findings.sort(key=lambda f: (f.line, f.rule))
    return findings

File: npm/src/rules/registry.js
// @ts-check
/**
 * 규칙 등록부. 규칙 하나는 파일 하나이고 `name` 과 `run(doc, config)` 을 내보낸다. 파이썬은 폴더를 걸어 찾고
 * 여기는 목록으로 든다. test/rules.test.js 가 목록과 폴더와 ruleDocs.json 이 같은지 본다.
 * 기술서 (왜, 어디서, 고치기, 안 잡는 것) 는 파이썬 docstring 의 투영 data/ruleDocs.json 에서 읽는다.
 */
import { enabled } from "../config/settings.js";
import { loadRuleCategories, loadRuleDocs } from "../data/load.js";
import * as countMismatch from "./document/countMismatch.js";
import * as enoughOnce from "./document/enoughOnce.js";
import * as fieldEcho from "./document/fieldEcho.js";
import * as keywordHeading from "./document/keywordHeading.js";
import * as keywordMissing from "./document/keywordMissing.js";
import * as noQuestion from "./document/noQuestion.js";
import * as promiseRecall from "./document/promiseRecall.js";
import * as tableOddCell from "./document/tableOddCell.js";
import * as duplicateBlock from "./code/duplicateBlock.js";
import * as firstResultDistance from "./code/firstResultDistance.js";
import * as inputFileSource from "./code/inputFileSource.js";
import * as installImport from "./code/installImport.js";
import * as platformApi from "./code/platformApi.js";
import * as confusable from "./orthography/confusable.js";
import * as spacing from "./orthography/spacing.js";
import * as spelling from "./orthography/spelling.js";
import * as factListParagraph from "./paragraph/factListParagraph.js";
import * as paraFragment from "./paragraph/paraFragment.js";
import * as topicBreak from "./paragraph/topicBreak.js";
import * as cliche from "./sentence/cliche.js";
import * as hardWord from "./sentence/hardWord.js";
import * as connectorRepeat from "./sentence/connectorRepeat.js";
import * as danglingDeixis from "./sentence/danglingDeixis.js";
import * as draftHistory from "./sentence/draftHistory.js";
import * as numberOrphan from "./sentence/numberOrphan.js";
import * as dash from "./sentence/dash.js";
import * as deixis from "./sentence/deixis.js";
import * as doubleNegative from "./sentence/doubleNegative.js";
import * as doublePassive from "./sentence/doublePassive.js";
import * as ellipsis from "./sentence/ellipsis.js";
import * as endingRepeat from "./sentence/endingRepeat.js";
import * as euiChain from "./sentence/euiChain.js";
import * as fillerOpener from "./sentence/fillerOpener.js";
import * as imperativePeriod from "./sentence/imperativePeriod.js";
import * as japaneseLoan from "./sentence/japaneseLoan.js";
import * as longSentence from "./sentence/longSentence.js";
import * as negationRedefine from "./sentence/negationRedefine.js";
import * as nounPile from "./sentence/nounPile.js";
import * as redundantPair from "./sentence/redundantPair.js";
import * as translationese from "./sentence/translationese.js";
import * as blockUnread from "./structure/blockUnread.js";
import * as bridgeRepeat from "./structure/bridgeRepeat.js";
import * as emojiBullet from "./structure/emojiBullet.js";
import * as headingQuestion from "./structure/headingQuestion.js";
import * as introImage from "./structure/introImage.js";
import * as loneSubheading from "./structure/loneSubheading.js";
import * as moreLater from "./structure/moreLater.js";
import * as headingSentence from "./structure/headingSentence.js";
import * as headingSkip from "./structure/headingSkip.js";
import * as headingUniform from "./structure/headingUniform.js";
import * as introLong from "./structure/introLong.js";
import * as sectionNoProse from "./structure/sectionNoProse.js";
import * as sectionResult from "./structure/sectionResult.js";

/**
 * @typedef {object} Rule
 * @property {string} name
 * @property {string} mechanism 세는 방법. MECHANISMS 의 키 하나
 * @property {(doc: import("../fingerprint/build.js").DocumentPrint, config: import("../config/settings.js").Config) => import("./finding.js").Finding[]} run
 */

/** @type {Rule[]} */
export const RULES = [
  cliche,
  connectorRepeat,
  danglingDeixis,
  dash,
  draftHistory,
  numberOrphan,
  deixis,
  doubleNegative,
  doublePassive,
  ellipsis,
  endingRepeat,
  euiChain,
  fillerOpener,
  hardWord,
  imperativePeriod,
  japaneseLoan,
  longSentence,
  negationRedefine,
  nounPile,
  redundantPair,
  translationese,
  factListParagraph,
  paraFragment,
  topicBreak,
  blockUnread,
  bridgeRepeat,
  emojiBullet,
  headingQuestion,
  introImage,
  loneSubheading,
  moreLater,
  headingSentence,
  headingSkip,
  headingUniform,
  introLong,
  sectionNoProse,
  sectionResult,
  countMismatch,
  enoughOnce,
  fieldEcho,
  tableOddCell,
  keywordHeading,
  keywordMissing,
  noQuestion,
  promiseRecall,
  spelling,
  spacing,
  confusable,
  inputFileSource,
  installImport,
  duplicateBlock,
  firstResultDistance,
  platformApi,
];

/**
 * 규칙이 세는 방법의 닫힌 집합. 파이썬 rules/registry.py 의 MECHANISMS 와 같은 값과 같은 순서다.
 * @type {Record<string, string>}
 */
export const MECHANISMS = {
  dictionary: "사전. 낱말 목록과 정규식이 맞는 자리",
  repeat: "반복. 같은 모양이 창 안에서 N 번",
  threshold: "셈. 지문의 값이 임계를 넘거나 모양이 계약과 다른 자리",
  contrast: "대조. 두 자리를 맞대 어긋난 곳",
  reader: "독자 상태. 문장 순서대로 독자가 손에 든 것과 본 것에 그 자리가 요구하는 것을 맞댄다",
};
for (const rule of RULES) {
  if (!(rule.mechanism in MECHANISMS)) {
    throw new Error(`규칙 ${rule.name} 의 기제 ${rule.mechanism} 은 닫힌 집합 밖이다 (${Object.keys(MECHANISMS).join(", ")}). 새 기제는 규칙 하나 때문에 들어오지 않는다. 멈춰서 묻는다`);
  }
}

const BY_NAME = new Map(RULES.map((rule) => [rule.name, rule]));

/** 이름 순. */
export function ruleNames() {
  return [...BY_NAME.keys()].sort();
}

/** @param {string} name */
export function ruleDoc(name) {
  const docs = loadRuleDocs();
  if (!(name in docs) || !BY_NAME.has(name)) throw new Error(`모르는 규칙: ${name}. hanlint rules 로 목록을 본다`);
  return docs[name];
}

/** @param {string} name */
export function ruleSummary(name) {
  return ruleDoc(name).split("\n")[0];
}

/**
 * 부류의 사람 이름. 파이썬 rules/registry.py 의 CATEGORY_TITLES 와 같은 값과 같은 순서다.
 * @type {Record<string, string>}
 */
export const CATEGORY_TITLES = {
  sentence: "문장 안에서 세는 것",
  paragraph: "문단 사이에서 세는 것",
  structure: "글의 짜임에서 세는 것",
  document: "두 자리를 대조해 세는 것",
  orthography: "표기와 띄어쓰기",
  code: "코드 블록 사이를 대조하는 것",
};

/** 규칙이 세는 방법. @param {string} name */
export function ruleMechanism(name) {
  const rule = BY_NAME.get(name);
  if (!rule) throw new Error(`모르는 규칙: ${name}. hanlint rules 로 목록을 본다`);
  return rule.mechanism;
}

/** 규칙이 사는 부류. 정본은 파이썬 쪽 폴더이고 여기는 그 투영을 읽는다. @param {string} name */
export function ruleCategory(name) {
  const categories = loadRuleCategories();
  if (!(name in categories)) throw new Error(`모르는 규칙: ${name}. hanlint rules 로 목록을 본다`);
  return categories[name];
}

/**
 * 인라인 제어가 그 줄에서 그 규칙을 껐는가. `*` 는 전부다.
 * @param {import("./finding.js").Finding} f
 * @param {[string, number, number][]} disabled
 */
export function isDisabledAt(f, disabled) {
  return disabled.some(([name, start, end]) => (name === "*" || name === f.rule) && start <= f.line && f.line <= end);
}

/**
 * @param {import("../fingerprint/build.js").DocumentPrint} doc
 * @param {import("../config/settings.js").Config} config
 * @returns {import("./finding.js").Finding[]}
 */
export function runAll(doc, config) {
  /** @type {import("./finding.js").Finding[]} */
  const findings = [];
  for (const name of ruleNames()) {
    if (!enabled(config, name)) continue;
    const rule = /** @type {Rule} */ (BY_NAME.get(name));
    for (const f of rule.run(doc, config)) if (!isDisabledAt(f, doc.disabled)) findings.push(f);
  }
  findings.sort((a, b) => a.line - b.line || (a.rule < b.rule ? -1 : a.rule > b.rule ? 1 : 0));
  return findings;
}
